### ISSUE TITLE
Automate membership of mailing lists

### ISSUE THREAD

**Issue by alysbrett** [ScottishCovidResponse/SCRCIssueTracking#25]
We have a private repository in this Github org with lists (TOML format) of people (Githubids, names, email addresses) who are part of SCRC and then lists of the membership of the individual projects and groupings within the coalition. We would like to use this as the source data to automate keeping the membership of various mailing lists up to date.

The mailing lists are managed by the University of Glasgow using Sympa and Richard Reeve has established that they can pull data from a database to manage their membership. His contact (Mark Meenan) says:

"Assuming a mysql/mariadb database
What I would need is an account to query the database - from lists.cent.gla.ac.uk - 130.209.16.184
and what should go into the sql statement for each query"

So we have permission to create an (ideally) MySQL database that contains tables with the email addresses from each group in our personnel lists. If we can give an SQL query for each mailing list we want to set up that just returns the members of the team (and read permission to the server), then I think they can just set up a cron job to do it every night (or whatever). They are actually very keen on this, as they hate manual updating of mailing lists, so they'll definitely help if need be.

This task is to set up a way that the mailing list managers can run SQL queries to get the up to date membership of various groups, with this data changing whenever a new version of our personnel files is commited within Github. If we need to host a database it may be possible to do this at STFC and Andrew Lahiff is the contact to ask about that.

**Comment by jholloc** [comment]
I've created a database schema: https://dbdiagram.io/d/5eaaa48f39d18f5553fe7920.

**Comment by nathan-cummings-ukaea** [comment]
With that schema and the .toml files I'll try out setting up a database on my machine with mariaDB so it should be easy to get it established on the VM at STFC once that is up and running. Andrew has already gotten the ball rolling with that.

**Comment by alysbrett** [comment]
Thanks Jonathan. I think a query getting the overall list of people and another getting the people assigned to the individual projects in the projects sub-folder would be a good start. The whole roles and skills thing might not be well-defined enough yet (I think there are a couple of concepts going on and not sure we've ironed out how we're hoping to use them). I think if you add a projects table and the ability to link people to projects that would be useful. Maybe people have skills and also roles on a project but may need some discussion and not sure we need that for the immediate mailing list purpose. @richardreeve may wish to comment...

**Comment by alysbrett** [comment]
Just a query to get the the overall list of people and email addresses for the full mailing list would be a good first step in fact. Should check if they can easily do diffs and update only changes at their end or if they are hoping to get lists of people to add and remove (in which case we'd need to store times of changes). Hope it's the first one!

**Comment by alysbrett** [comment]
Confirmation that we want a mailing list per file in the projects sub-folder and one overall with everyone. I may have broken any automated access to the files as we have just been sorting out permissions for all the Github repos, ie removing global permissions and adding specific ones back as needed. I gave the people  on this issue read access to personnel but not sure what you will need for the automation so let me know.

**Comment by alysbrett** [comment]
(not sure why it says I unassigned @alahiff when I added that comment...)

**Comment by jholloc** [comment]
@alysbrett we will need to access the personnel repo from the STFC VM, any thoughts? I can add the VM's SSH key to my account for now.

**Comment by alahiff** [comment]
@jholloc Can you create a token in GitHub which you can use on the VM to access the repo? I haven't actually used tokens in GitHub before but it may be possible to do things like this...

**Comment by alysbrett** [comment]
I can add a deploy key (https://developer.github.com/v3/guides/managing-deploy-keys/#deploy-keys) in that repo if you like?

I just need a name for it and the public SSH key

**Comment by jholloc** [comment]
Server public key:

`ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDMyKnJkyrCTZ/rCnyvWVr/yYtlqs2NsLUpX3Ybcfb6jogi1igfadzSutB/QnSrfaPhovBzyr+gFFPZ0S5QJCA3CuWY/ktFVxDg1JNwuoJi/7kje+MD4/qXrC+FfdEYbXa5YLlZ4lFXP6vDRIi21opX5C9Yre2IHG3eQIAr6nZgy+8HYNuI3NdVhjAVul9YkMAs8RQYm1KR6FJ1fCcPLL6s+ekmEtO2pUUiQ3ohiO8wyfAGLy6TavjXStQmg7OpcxjeqF6/Ww/pUzd4xo8jkGDTAIodjbam8SRV19ncWQJb7Sg0iSjsnVRiWOOC+s2DId8D0A6SQlPk9XHnXxPeUxKx dbadmin@covid-mariadb-prod`

The name could be something like personnel-db

**Comment by alysbrett** [comment]
Deploy key added.

**Comment by jholloc** [comment]
Thanks, that works.

**Comment by nathan-cummings-ukaea** [comment]
`SELECT u.email FROM user AS u, user_projects AS up, project AS p WHERE up.user_id = u.id AND up.project_id = p.id AND p.name = 'BICI';`

^ gives all email addresses assigned to project BICI

I'm setting up a cronjob to update the database which will run at 9pm daily.

**Comment by alysbrett** [comment]
Great! Maybe time to loop in the people at Glasgow on the mailing list end - @richardreeve should be able to put them in touch next week

**Comment by richardreeve** [comment]
Yes, I'll do that now while I see this. Sorry I've been a bit hopeless responding - I've only just seen this through the millions of notifications I'm getting from all of the incoming issues.

**Comment by nathan-cummings-ukaea** [comment]
Port 3306 on 130.246.213.193 is open for their IP, the database has user called 'readonly' with password 'readonly'. I need to change a couple of permissions settings before they'll be able access the database but I'll probably have that sorted this afternoon.

That should be all the info needed. Let me know if I've missed anything.

**Comment by richardreeve** [comment]
Brilliant. I'm just waiting on confirmation of the list names from the list owners, but apart from that I've got the email ready to go on Monday morning.

In the meanwhile I've updated the personnel projects folder with some updated names and some new projects (sorry for that), and this is what I was planning on sending Mark:

> We currently need 11 mailing lists, but we will add more once we have sorted things out). They will all managed by myself, Lynn and Alys for now (if you can have a manager outside Glasgow? If not, it doesn't matter).
> 
> scrc-all (Nathan, this will be the complete list from people.toml)
> scrc-management (this is projects/management.toml)
> 
> Then 8 lists of individual groups of developers:
> scrc-simple-network-sim (projects/python-sim.toml)
> scrc-contact-tracing-model (projects/java-sim-sm.toml)
> scrc-covid-simulation-model (projects/java-sim-pb.toml)
> scrc-covid19-eeramodel (projects/cpp-inf-cp.toml)
> scrc-simulation-jl (projects/julia-sim.toml)
> scrc-coronabici (projects/cpp-inf-cp.toml)
> scrc-data (I've just added projects/data.toml)
> scrc-visualisation (ditto projects/visualisation.toml)
> 
> Port 3306 on 130.246.213.193 is open for your IP, the database has user called 'readonly' with password 'readonly'.
> 
> And finally one list that contains all of the 8 developer lists above (not -all or -management), and doesn't need a database to back it so long as it will update whenever its children do:
> 
> scrc-developers

I can adjust that to give the exact SQL statements, but I just wanted to check:

1. How would Mark get all of the names for scrc-all?
2. is `p.name` in the SQL equal to the `name = "xxx"` in the toml file for the other projects?

**Comment by nathan-cummings-ukaea** [comment]
1. `SELECT user.full_name FROM user;` will give the full name of everyone.
    `SELECT user.email FROM user;` will give their email addresses.

2. It is indeed. Since some of the project names have now changed I refreshed the db so it doesn't have the projects duplicated - to avoid confusion. If project names are likely to change further let me know, I'll need to make the script that updates the database slightly smarter to remove projects that no longer exist in the .toml files if that is the case.

If you want a management list, making management a project would be useful, for now that .toml is ignored.

**Comment by jholloc** [comment]
@nathan-cummings-ukaea I'd be tempted to just delete everything in the database before the update script is run... it's only going to be run once a day and for a few dozen users.

**Comment by nathan-cummings-ukaea** [comment]
I will do if it becomes an issue. For now it's all set up, the cronjob is done so all being well everything should work for Mark. Let me know if anything goes wrong.

**Comment by richardreeve** [comment]
Thanks. Management does need a list, so I've given it a name entry - "Management" - to (hopefully) simplify things and generally reformatted it to the same as the other projects. Can you add it back in, please? For now, we'll hope not to change the project names.

We may well need to add in more projects though... can you just read anything in the folder? Or is that what you do in fact?

And we'll raise a new issue if we want to add in role-based lists.

The final thing is, would it be possible to check the cron job script into this repository along with a brief description of how it was set up in case you're not available (maybe also the schema if that's possible)?

Hopefully this is then right:

> The server is now up and running:
> 
> Port 3306 on 130.246.213.193 is open for your IP, the database has a user called 'readonly' with password 'readonly'.
> 
> We currently need 11 mailing lists, but we will add more once we have sorted things out). They will all managed by myself, Lynn and Alys for now (if you can have a manager outside Glasgow? If not, it doesn't matter). First the list for everyone:
> 
> scrc-all (to access this you need "SELECT user.email FROM user;")
> 
> For all of the rest you need "SELECT u.email FROM user AS u, user_projects AS up, project AS p WHERE up.user_id = u.id AND up.project_id = p.id AND p.name = 'PROJECTNAME'"
> 
> where 'PROJECTNAME' is replaced by the text below in brackets:
> 
> scrc-exec ('PROJECTNAME' is 'Management')
> 
> Then 8 lists of individual groups of developers:
> scrc-simple-network-sim ('Simple Network Sim')
> scrc-contact-tracing-model ('Contact Tracing Model')
> scrc-covid-simulation-model ('Covid Simulation Model')
> scrc-covid19-eeramodel ('COVID-19 EERAModel')
> scrc-simulation-julia ('Simulation.jl')
> scrc-coronabici ('CoronaBICI')
> scrc-data ('Data')
> scrc-vis ('Visualisation')
> 
> And finally one list that contains all of the 8 developer lists above (not -all or -exec), and doesn't need a database to back it so long as it will update whenever its children do:
> 
> scrc-developers
>

**Comment by nathan-cummings-ukaea** [comment]
Yep, I’ll re-add management either later today or tomorrow (before Monday for sure).

Yep, new projects should be picked up automatically, it does indeed iterate through the folder.

The shell script that the cron job runs is literally just the git pull then run the python script that Jonathan wrote. That, along with the schema could certainly be uploaded if you wish.

**Comment by richardreeve** [comment]
Yes, sorry - I meant the python script. Thanks.

**Comment by nathan-cummings-ukaea** [comment]
Hi @richardreeve, re-including management.toml is throwing an Exception right now as there are people in there that aren't in 'people.toml'. Once they are added let me know and I can make sure the 'management' project gets generated. Alternatively I could ignore people in 'management.toml' that aren't in 'people.toml' but that may lead to someone being missed off a mailing list.

**Comment by richardreeve** [comment]
That should be fixed now - I just didn't update someone's github username there when they gave it to me. In general I think it would be good to ignore those errors anywhere in building the mailing lists, so they do get mostly updated even if usernames are missing.

However, is there's any way of messaging one of us to tell us there's a problem so we can fix it?

If there's no easy way for the server script to report back, then instead it might be easier for me to add some CI in the repository so that any changes to the repo checks that all github usernames are referenced in people.toml? That might be more sensible.

**Comment by nathan-cummings-ukaea** [comment]
Cheers, the db has a management project now - looks good.

The CI approach would be more robust, and would be the best adherence to 'failing early' I guess. In the mean time the logs for the cron job will at least let us know if the update failed the night before.

Would you like the python and mysql script in the personnel repo or elsewhere?

**Comment by richardreeve** [comment]
Yes, good point. I’ll add an issue to remind me. And can you put the code somewhere in the personnel repo I think, so it’s all together.

**Comment by nathan-cummings-ukaea** [comment]
Could I have permission to push to the personnel repo so I can add the db scripts?

**Comment by alysbrett** [comment]
You should be able to write to it now @nathan-cummings-ukaea

**Comment by alysbrett** [comment]
From email exchanges today I think this can be closed once @nathan-cummings-ukaea confirms he has put the relevant scripts in the repo etc and we see that new people added this evening have shown up correctly. When you have finished this work and could look at the next thing Nathan, there is the related matter of #74 if you are familiar with (or willing to experiment with) Github CI.

**Comment by nathan-cummings-ukaea** [comment]
The relevant scripts are indeed now in the personnel repo (thanks to @jholloc for putting these together). Checked the logs this morning, the update is hourly and running successfully.

I've not used Github CI before but happy to give it a go. Feel free to assign me to this task.

**Comment by alysbrett** [comment]
Think this is now done then and remaining work is in the CI related issue... thanks very much @jholloc @nathan-cummings-ukaea  and @alahiff !
